### PR TITLE
Merge MARL-new into master

### PR DESCRIPTION
**Pull request by webepasc**

### CODE REVIEW

File: source/modules/solver/agent/agent.cpp [hunk 1]
```diff
@@ -679,162 +684,297 @@ void Agent::processEpisode(knlohmann::json &episode)
...
+
+  /* Creating a selection of unique expIds, agentIds
+   * Important: this assumes the minibatch ids are sorted.
+   */
+
+  // Create Buffers
   std::vector<size_t> updateBatch;
-  updateBatch.push_back(0);
-  for (size_t i = 1; i < miniBatchSize; i++)
-    if (miniBatch[i] != miniBatch[i - 1]) updateBatch.push_back(i);
```
Comment (wadaniel) [inline source/modules/solver/agent/agent.cpp:715]: cant we keep this logic and check expId and agentId in the tuple? instead of the for loop + while loop below, the miniBatch is sorted. In the case "Baseline" it will work if we sorted the minibatch by expId and then agentId as 2nd criteria. In the case "Experience" it will work too?

File: source/modules/solver/agent/agent.cpp [hunk 2]
```diff
@@ -679,162 +684,297 @@ void Agent::processEpisode(knlohmann::json &episode)
...
-                              : std::transform(omp_out.begin(), omp_out.end(), omp_in.begin(), omp_out.begin(), std::plus <int>())) \
-  initializer(omp_priv = decltype(omp_orig)(omp_orig.size()))
+  // Fill updateMinibatch and updatePolicyData
+  size_t a;
+  for (size_t b = 0; b < miniBatchSize; b += a)
+  {
+    // Add new unique combination
+    updateBatch.push_back(b);
+    updateMinibatch.push_back(miniBatch[b]);
+    updatePolicyData.push_back(policyData[b]);
```
Comment (wadaniel) [inline source/modules/solver/agent/agent.cpp:798]: shouldnt it be policyData[miniBatch[b]]?

File: source/modules/solver/agent/agent.cpp [hunk 3]
```diff
@@ -679,162 +684,297 @@ void Agent::processEpisode(knlohmann::json &episode)
...
-#pragma omp declare reduction(vec_int_plus                                                                                          \
-                              : std::vector <int>                                                                                   \
-                              : std::transform(omp_out.begin(), omp_out.end(), omp_in.begin(), omp_out.begin(), std::plus <int>())) \
-  initializer(omp_priv = decltype(omp_orig)(omp_orig.size()))
+  // Fill updateMinibatch and updatePolicyData
+  size_t a;
+  for (size_t b = 0; b < miniBatchSize; b += a)
+  {
+    // Add new unique combination
+    updateBatch.push_back(b);
```
Comment (wadaniel) [inline source/modules/solver/agent/agent.cpp:796]: is the updateBatch still needed?

Comment (wadaniel) [reply source/modules/solver/agent/agent.cpp:796]: if we use updateMiniBatch everywhere whereever we had miniBatch[updateBatch[b]]

File: source/modules/solver/agent/agent.cpp [hunk 4]
```diff
@@ -679,162 +684,297 @@ void Agent::processEpisode(knlohmann::json &episode)
...
+  /* Now taking care of advances correlation features for MARL */
+
+  if (_multiAgentCorrelation)
+  {
+    #pragma omp parallel for reduction(vec_int_plus : offPolicyCountDelta) schedule(guided, numAgents)
+    for( size_t i = 0; i<updateBatch.size(); i++ )
     {
-      const double logCutOff = (double)_problem->_agentsPerEnvironment * std::log(_experienceReplayOffPolicyCurrentCutoff);
+      const size_t batchId = updateBatch[i];
+      const size_t expId = miniBatch[batchId].first;
```
Comment (wadaniel) [inline source/modules/solver/agent/agent.cpp:894]: e.g. here we can use updateMiniBatch

File: source/modules/solver/agent/agent.cpp [hunk 5]
```diff
@@ -679,162 +684,297 @@ void Agent::processEpisode(knlohmann::json &episode)
...
-          truncStateValue[d] = calculateStateValue(expTruncatedStateSequence);
-        else
-          truncStateValue[d] = calculateStateValue(expTruncatedStateSequence, d);
-
-        // Get value of trucated state
-        if (std::isfinite(truncStateValue[d]) == false)
-          KORALI_LOG_ERROR("Calculated state value for truncated state returned an invalid value: %f\n", truncStateValue[d]);
-      }
+      const size_t batchId = updateBatch[i];
+      const size_t expId = miniBatch[batchId].first;
```
Comment (wadaniel) [inline source/modules/solver/agent/agent.cpp:945]: e.g. here we can use updateMiniBatch

**Review by wadaniel — APPROVED** [review]

File: source/modules/solver/agent/agent.config [hunk 6]
```diff
@@ -155,20 +155,25 @@
...
-    "Name": [ "Strong Truncation Variant" ],
-    "Type": "bool",
-    "Description": "Only important in strong correlation case, if set to True we first multiplicate and then truncate, if set to False we first truncate and then multiplicate."
+    "Name": [ "Multi Agent Sampling" ],
+    "Type": "std::string",
+    "Options": [
+        { "Value": "Tuple", "Description": "Sample expId and use same value for all agents." },
+        { "Value": "Baseline", "Description": "Sample expId and agentId." },
+        { "Value": "Experience", "Description": "Sample expId separately for all agents." }
+    ],
```
[the remark this answers is not in the document]
Comment (webepasc) [reply source/modules/solver/agent/agent.config:176]: I removed this

File: source/modules/solver/agent/agent.cpp [hunk 7]
```diff
@@ -679,162 +684,297 @@ void Agent::processEpisode(knlohmann::json &episode)
...
+        miniBatch[b * numAgents + a].second = std::floor(ax * (float)(numAgents - 1));
+      }
+    }
   }
 
   // Sorting minibatch -- this helps with locality and also
   // to quickly detect duplicates when updating metadata
-  std::sort(miniBatch.begin(), miniBatch.end());
+  std::sort(miniBatch.begin(), miniBatch.end(), secondSmaller);
+  std::stable_sort(miniBatch.begin(), miniBatch.end());
```
[the remark this answers is not in the document]
Comment (webepasc) [reply source/modules/solver/agent/agent.cpp]: also fixed, std::sort on tuple does the wanted sorting of first and then second element

File: source/modules/solver/agent/agent.cpp (hunk 1 above)
Comment (webepasc) [reply source/modules/solver/agent/agent.cpp:612]: This is nicyfied

File: tests/unit/modules/solver/deepSupervisor.cpp [hunk 8]
```diff
@@ -192,21 +192,6 @@ namespace
    supervisorJs["Loss Function"] = "Direct Gradient";
    ASSERT_NO_THROW(supervisor->setConfiguration(supervisorJs));
 
-   supervisorJs = baseOptJs;
```
Comment (wadaniel) [inline tests/unit/modules/solver/deepSupervisor.cpp:195]: why is this removed?

File: tests/unit/modules/solver/agent.cpp [hunk 9]
```diff
@@ -1518,26 +1370,6 @@ namespace
 
    // Covering variable functions (no effect)
    ASSERT_NO_THROW(a->applyVariableDefaults());
-
```
Comment (wadaniel) [inline tests/unit/modules/solver/agent.cpp:1521]: why did we remove this?

File: tests/unit/modules/solver/agent.cpp [hunk 10]
```diff
@@ -1610,36 +1442,6 @@ namespace
 
     // Covering variable functions (no effect)
     ASSERT_NO_THROW(a->applyVariableDefaults());
-
```
Comment (wadaniel) [inline tests/unit/modules/solver/agent.cpp:1613]: this

File: source/modules/problem/reinforcementLearning/reinforcementLearning.cpp.base [hunk 11]
```diff
@@ -67,17 +69,19 @@ void __environmentWrapper()
   (*agent)["Launch Id"] = _launchId++;
   agent->run(__envFunctionId);
 
-  // If this is not the leader rank within the worker group, return immediately without checking termination state
-  if (_conduit->isWorkerLeadRank() == false) return;
```
[the remark this answers is not in the document]
Comment (webepasc) [reply source/modules/problem/reinforcementLearning/reinforcementLearning.cpp.base:71]: Because it was a bug

File: source/modules/problem/reinforcementLearning/reinforcementLearning.cpp.base [hunk 12]
```diff
@@ -180,7 +175,10 @@ void __className__::runTrainingEpisode(Sample &agent)
     // Checking if we requested the given number of actions in between policy updates and it is not a terminal state
     if ((_actionsBetweenPolicyUpdates > 0) &&
         (agent["Termination"] == "Non Terminal") &&
-        (actionCount % _actionsBetweenPolicyUpdates == 0)) requestNewPolicy(agent);
+        (actionCount % _actionsBetweenPolicyUpdates == 0))
```
[the remark this answers is not in the document]
Comment (webepasc) [reply source/modules/problem/reinforcementLearning/reinforcementLearning.cpp.base:178]: There is some examples that update the policy after >1 action

File: source/modules/solver/agent/agent.hpp.base [hunk 13]
```diff
@@ -131,62 +136,67 @@ class __className__ : public __parentClassName__
...
    * @brief Contains the policy information produced at the moment of the action was taken
    */
-  cBuffer<policy_t> _expPolicyBuffer;
+  cBuffer<std::vector<policy_t>> _expPolicyBuffer;
 
   /**
    * @brief Indicates whether the experience is on policy, given the specified off-policiness criteria
    */
-  cBuffer<bool> _isOnPolicyBuffer;
+  cBuffer<std::vector<char>> _isOnPolicyBuffer;
```
[the remark this answers is not in the document]
Comment (webepasc) [reply source/modules/solver/agent/agent.hpp.base:164]: this gives a bug, because vector<bool> is not thread save

File: tests/unit/modules/solver/deepSupervisor.cpp (hunk 8 above)
Comment (webepasc) [reply tests/unit/modules/solver/deepSupervisor.cpp:195]: in deep learning, each minibatch is forwarded once and we go with the next iteration after an epoch is completed, this is something that was never used

File: tests/unit/modules/solver/agent.cpp (hunk 9 above)
Comment (webepasc) [reply tests/unit/modules/solver/agent.cpp:1521]: same reason as above

File: tests/unit/modules/solver/agent.cpp (hunk 10 above)
Comment (webepasc) [reply tests/unit/modules/solver/agent.cpp:1613]: This was replaced and was not used anymore

File: source/modules/solver/agent/agent.cpp.base [hunk 14]
```diff
@@ -280,58 +313,62 @@ void __className__::trainingGeneration()
...
+    {
+      _stateRescalingMeans[a][d] = sumStates[a][d] / (float)_stateBuffer.size();
+      if (std::isfinite(_stateRescalingMeans[a][d]) == false) _stateRescalingMeans[a][d] = 0.0f;
 
-    _stateRescalingSigmas[d] = std::sqrt(squaredSumStates[d] / (float)_stateBuffer.size() - _stateRescalingMeans[d] * _stateRescalingMeans[d]);
-    if (std::isfinite(_stateRescalingSigmas[d]) == false) _stateRescalingSigmas[d] = 1.0f;
-    if (_stateRescalingSigmas[d] <= 1e-9) _stateRescalingSigmas[d] = 1.0f;
+      _stateRescalingSigmas[a][d] = std::sqrt(squaredSumStates[a][d] / (float)_stateBuffer.size() - _stateRescalingMeans[a][d] * _stateRescalingMeans[a][d]);
+      if (std::isfinite(_stateRescalingSigmas[a][d]) == false) _stateRescalingSigmas[a][d] = 1.0f;
+      if (_stateRescalingSigmas[a][d] <= 1e-9) _stateRescalingSigmas[a][d] = 1.0f;
```
[the remark this answers is not in the document]
Comment (webepasc) [reply source/modules/solver/agent/agent.cpp.base:362]: well, actually something is wrong with the estimator, it would fail before if the state is non-finite. Since we do state-rescaling once, I would suggest to replace this estimator by a safe one...

File: source/modules/solver/deepSupervisor/deepSupervisor.cpp.base [hunk 15]
```diff
@@ -188,6 +320,124 @@ std::vector<std::vector<float>> &__className__::getEvaluation(const std::vector<
...
+  auto nnHyperparameters = KORALI_GET(std::vector<float>, sample, "Hyperparameters");
+  _neuralNetwork->setHyperparameters(nnHyperparameters);
+  sample._js.getJson().erase("Hyperparameters");
+
+  // Getting input from sample
+  auto inputDataFlat = KORALI_GET(std::vector<float>, sample, "Input Data");
+  sample._js.getJson().erase("Input Data");
+
+  // Getting input dimensions
+  auto inputDims = KORALI_GET(std::vector<size_t>, sample, "Input Dims");
```
[the remark this answers is not in the document]
Comment (webepasc) [reply source/modules/solver/deepSupervisor/deepSupervisor.cpp.base:424]: fixed

File: source/modules/solver/deepSupervisor/deepSupervisor.cpp.base [hunk 16]
```diff
@@ -188,6 +320,124 @@ std::vector<std::vector<float>> &__className__::getEvaluation(const std::vector<
...
+  auto inputDataFlat = KORALI_GET(std::vector<float>, sample, "Input Data");
+  sample._js.getJson().erase("Input Data");
+
+  // Getting solution from sample
+  auto solutionDataFlat = KORALI_GET(std::vector<float>, sample, "Solution Data");
+  sample._js.getJson().erase("Solution Data");
+
+  // Getting input dimensions
+  auto inputDims = KORALI_GET(std::vector<size_t>, sample, "Input Dims");
+  size_t N = inputDims[0];
```
[the remark this answers is not in the document]
Comment (webepasc) [reply source/modules/solver/deepSupervisor/deepSupervisor.cpp.base:364]: fixed

File: source/modules/solver/deepSupervisor/deepSupervisor.cpp.base [hunk 17]
```diff
@@ -98,68 +114,184 @@ void __className__::initialize()
...
-          _currentLoss += gradientVector[b][i] * gradientVector[b][i];
-      _currentLoss = _currentLoss / ((float)N * 2.0f);
+  // Hyperparameter gradient storage
+  std::vector<float> nnHyperparameterGradients;
+
+  // In case we run Mean Squared Error with concurrency, distribute the work among samples
+  if (_lossFunction == "Mean Squared Error" && _batchConcurrency > 1)
+  {
+    // Calculating per worker dimensions
+    size_t NW = _problem->_trainingBatchSize / _batchConcurrency;
```
[the remark this answers is not in the document]
Comment (webepasc) [reply source/modules/solver/deepSupervisor/deepSupervisor.cpp.base:206]: fixed

File: source/modules/solver/deepSupervisor/deepSupervisor.cpp.base [hunk 18]
```diff
@@ -98,68 +114,184 @@ void __className__::initialize()
...
+  if (_problem->_testingBatchSize != _problem->_inputData.size())
+    KORALI_LOG_ERROR("Testing Batch size %lu different than that of input data (%lu).\n", _problem->_inputData.size(), _problem->_testingBatchSize);
+
+  // In case we run Mean Squared Error with concurrency, distribute the work among samples
+  if (_batchConcurrency > 1)
   {
-    // If we use an MSE loss function, we need to update the gradient vector with its difference with each of batch's last timestep of the NN output
-    if (_lossFunction == "Mean Squared Error")
+    // Calculating per worker dimensions
+    size_t NW = _problem->_testingBatchSize / _batchConcurrency;
```
[the remark this answers is not in the document]
Comment (webepasc) [reply source/modules/solver/deepSupervisor/deepSupervisor.cpp.base:134]: fixed

File: source/modules/solver/agent/continuous/VRACER/VRACER.cpp.base [hunk 19]
```diff
@@ -9,224 +9,299 @@
...
+      {
+        const float correlationFactor = _productImportanceWeightBuffer[expId] / _importanceWeightBuffer[expId][agentId];
+        for (size_t i = 0; i < polGrad.size(); i++)
+          polGrad[i] *= correlationFactor;
+      }
+
       // Set Gradient of Loss wrt Params
-      for (size_t i = 0; i < _policyParameterCount; i++)
-        gradientLoss[1 + i] = _experienceReplayOffPolicyREFERBeta * lossOffPolicy * polGrad[i];
+      for (size_t i = 0; i < 2 * _problem->_actionVectorSize; i++)
```
[the remark this answers is not in the document]
Comment (webepasc) [reply source/modules/solver/agent/continuous/VRACER/VRACER.cpp.base:216]: fixed

File: source/modules/solver/agent/continuous/VRACER/VRACER.cpp.base [hunk 20]
```diff
@@ -9,224 +9,299 @@
...
   // Obtaining Minibatch experience ids
-  const auto miniBatch = generateMiniBatch(_miniBatchSize);
+  const auto miniBatch = generateMiniBatch();
 
   // Gathering state sequences for selected minibatch
-  const auto stateSequence = getMiniBatchStateSequence(miniBatch);
+  const auto stateSequenceBatch = getMiniBatchStateSequence(miniBatch);
+
+  // Prepare stateSequenceBatch and miniBatch
+  auto _miniBatch = miniBatch;
```
[the remark this answers is not in the document]
Comment (webepasc) [reply source/modules/solver/agent/continuous/VRACER/VRACER.cpp.base:92]: fixed

File: source/modules/conduit/distributed/distributed.hpp.base [hunk 21]
```diff
@@ -72,6 +67,7 @@ class __className__ : public __parentClassName__
   int getRootRank();
   bool isRoot() override;
   bool isWorkerLeadRank() override;
+  size_t getWorkerCount() override;
```
[the remark this answers is not in the document]
Comment (webepasc) [reply source/modules/conduit/distributed/distributed.hpp.base:70]: fixed

File: source/modules/conduit/distributed/distributed.cpp.base [hunk 22]
```diff
@@ -129,12 +146,12 @@ void __className__::broadcastMessageToWorkers(knlohmann::json &message)
   // Run broadcast only if this is the master process
   if (!isRoot()) return;
 
-  string messageString = message.dump();
-  size_t messageStringSize = messageString.size();
+  // Serializing message in binary form
+  std::vector<std::uint8_t> msgData = knlohmann::json::to_cbor(message);
```
[the remark this answers is not in the document]
Comment (webepasc) [reply source/modules/conduit/distributed/distributed.cpp.base:150]: fixed

File: source/modules/conduit/distributed/distributed.cpp.base [hunk 23]
```diff
@@ -108,14 +119,20 @@ void __className__::terminateServer()
   auto terminationJs = knlohmann::json();
   terminationJs["Conduit Action"] = "Terminate";
 
-  string terminationString = terminationJs.dump();
-  size_t terminationStringSize = terminationString.size();
+  // Serializing message in binary form
+  std::vector<std::uint8_t> msgData = knlohmann::json::to_cbor(terminationJs);
```
[the remark this answers is not in the document]
Comment (webepasc) [reply source/modules/conduit/distributed/distributed.cpp.base:123]: fixed

File: source/modules/conduit/distributed/distributed.cpp.base [hunk 24]
```diff
@@ -46,12 +55,13 @@ void __className__::initialize()
...
-  while (!_workerQueue.empty()) _workerQueue.pop();
+  _workerQueue.clear();
 
   // Putting workers in the queue
-  for (int i = 0; i < _workerCount; i++)
-    _workerQueue.push(i);
+  for (int i = 0; i < _workerCount; i++) _workerQueue.push_back(i);
+
+  // Korali engine as default, is the n+1th worker
+  int curWorker = _workerCount + 1;
```
[the remark this answers is not in the document]
Comment (webepasc) [reply source/modules/conduit/distributed/distributed.cpp.base:64]: nope, this is modified further below

File: source/modules/conduit/distributed/distributed.cpp [hunk 25]
```diff
@@ -112,14 +123,20 @@ void Distributed::terminateServer()
   auto terminationJs = knlohmann::json();
   terminationJs["Conduit Action"] = "Terminate";
 
-  string terminationString = terminationJs.dump();
-  size_t terminationStringSize = terminationString.size();
+  // Serializing message in binary form
+  std::vector<std::uint8_t> msgData = knlohmann::json::to_cbor(terminationJs);
```
[the remark this answers is not in the document]
Comment (webepasc) [reply source/modules/conduit/distributed/distributed.cpp:127]: fixed

File: source/modules/conduit/distributed/distributed.cpp [hunk 26]
```diff
@@ -23,21 +23,30 @@ void Distributed::initialize()
...
+  // If this is a root rank, check whether configuration is correct
+  if (isRoot())
+  {
+    if (_rankCount < 2) KORALI_LOG_ERROR("Korali Distributed applications require at least 2 MPI ranks to run.\n");
+    if (_ranksPerWorker < 1) KORALI_LOG_ERROR("The distributed conduit requires that the ranks per worker is equal or larger than 1, provided: %d\n", _ranksPerWorker);
+    if (_engineRanks < 1) KORALI_LOG_ERROR("The distributed conduit requires that the engine ranks is equal or larger than 1, provided: %d\n", _engineRanks);
+    if (workerRemainder != 0) KORALI_LOG_ERROR("Korali was instantiated with %lu MPI ranks (minus %lu for the engine), divided into %lu workers. This setup does not provide a perfectly divisible distribution, and %lu unused ranks remain.\n", _rankCount, _engineRanks, _workerCount, workerRemainder);
+  }
 
+  // Syncrhonizing all ranks
```
[the remark this answers is not in the document]
Comment (webepasc) [reply source/modules/conduit/distributed/distributed.cpp:46]: fixed

File: examples/reinforcement.learning/cartpole/_model/env.py [hunk 27]
```diff
@@ -71,7 +73,7 @@ def multienv(s):
     s["Reward"] = cart.getReward() * 0.1
    
   # Storing New State
-  s["State"] = cart.getState().tolist()
+  s["State"] = (cart.getState()).tolist()
```
[the remark this answers is not in the document]
Comment (webepasc) [reply examples/reinforcement.learning/cartpole/_model/env.py:76]: fixed

File: examples/hierarchical.bayesian/extended/run-phase1.py [hunk 28]
```diff
@@ -64,7 +64,4 @@
   eList.append(e)
 
 k = korali.Engine()
-k["Conduit"]["Type"] = "Concurrent"
```
[the remark this answers is not in the document]
Comment (webepasc) [reply examples/hierarchical.bayesian/extended/run-phase1.py:67]: because it yields problems for the tests due to the use of pipes

File: examples/features/multiple.experiments/run-cmaes.py [hunk 29]
```diff
@@ -71,15 +71,13 @@
...
 
-# Running first 100 generations
+# Running first 10 generations
 k.run(eList)
 
-# Running next 100 generations
+# Running next 10 generations
 for e in eList:
-  e["Solver"]["Termination Criteria"]["Max Generations"] = 150
+  e["Solver"]["Termination Criteria"]["Max Generations"] = 10
```
[the remark this answers is not in the document]
Comment (webepasc) [reply examples/features/multiple.experiments/run-cmaes.py:82]: indeed, fixed

**Review by wadaniel — APPROVED** [review]
